### ISSUE TITLE
Why loss becomes zero after calling xm.optimizer_step()?

### ISSUE THREAD

**Issue by 1374839016** [pytorch/xla#2763]
## ❓ Questions and Help
I use pytorch xla to train my model on kaggle, but something strange is that ,the first epoch , I calculate loss with mesh_reduce function ,it return a correct value,but after calling  xm.optimizer_step() function ,the loss become zero,after the first epoch,the loss is zero in each opoch.
But if I delete optimizer_step() , the loss become correct in several opochs ,but still become zero after several epochs.
why this issue occured and what should I do to solve the issue?Thanks.
![image](https://user-images.githubusercontent.com/51019236/106374824-7eac6f80-63c1-11eb-880e-408e1eb317b8.png)
![image](https://user-images.githubusercontent.com/51019236/106374834-984db700-63c1-11eb-9937-aeb211d3d8ef.png)

**Comment by taylanbil** [comment]
It's hard to debug issues such as these w/o a runnable repro script. Nothing pops in the image above to me that would cause something like this. Does this happen w/ 1 core? Could you provide more info, and better, a script that I can run and reproduce this?

**Comment by 1374839016** [comment]
Yes,thanks for your help.It's a kaggle kernel,I made this kernel public,you can fork and run, and you can see the bug.
[https://www.kaggle.com/jianxinhu/efficientnet-5-fold-train-mian-baseline-add-tpu](url)

**Comment by taylanbil** [comment]
Does this happen with one core?

**Comment by 1374839016** [comment]
I follow your advice and rerun my code,but the error still happen even with a single TPU core.I try to use different sampler,if I use torch.utils.data.RandomSampler ,at first epoch,the loss is not zero,but since second epoch,the loss become zero again,which has the same phenomenon as 8 TPU cores.
![image](https://user-images.githubusercontent.com/51019236/106836675-9131ec00-66d4-11eb-9188-189c78142046.png)
But I also try to use distributed sampler,the original sampler (torch.utils.data.distributed.DistributedSampler),which I used when train my network with 8 TPU cores,the phenomenon is very strange,the loss was zero at first epoch,then it will become normal for several epochs,then return to zero.
![image](https://user-images.githubusercontent.com/51019236/106837680-e1f61480-66d5-11eb-8129-5bd6f9704a58.png)

**Comment by taylanbil** [comment]
running with one core, could you inspect your input, output and target when loss is zero? i feel like this could be a data related issue.

**Comment by 1374839016** [comment]
Yes,you are right.I print the output,at first mini-batch,the model has normal output value,but after that ,the model output nan and loss become zero,but I don't know why this happened.
`start_training
tensor([[ 0.3164,  3.7344,  6.4375,  8.5000, -1.2109],
        [ 1.9922,  0.8203,  3.9375,  1.1250,  0.5039],
        [ 1.0000,  0.6953,  5.1562,  6.9688, -1.6797],
        [-0.9492,  2.6250,  1.2031,  3.3125, -0.3652],
        [-1.4609, -2.5469, -0.7773,  1.8984, -1.0859],
        [ 1.9141,  0.0221,  2.6406,  1.4531, -1.0391],
        [-0.4961, -0.3125,  1.8516,  8.1250, -2.2188],
        [-2.8750,  2.6094,  4.6250, -0.1436,  0.2354],
        [ 4.0625, -0.3223,  8.3750,  7.2812, -2.7656],
        [-0.9961,  0.5938, -0.3145,  0.3613,  0.9492],
        [ 3.2188,  1.7109,  2.7812,  2.0469, -0.7070],
        [ 4.5312,  0.4824, -0.2012,  0.0381, -1.8828],
        [ 0.9414, -1.5469,  0.4902,  1.1484,  0.9023],
        [-0.7773,  0.1934,  1.4297,  0.3105, -1.4141],
        [ 1.6875,  0.4102,  3.6719,  1.5703, -1.7969],
        [ 0.6836,  1.7969,  1.7812,  2.0469,  1.7812]], device='xla:1',
       grad_fn=<AddmmBackward>) tensor([3, 2, 3, 1, 3, 2, 3, 3, 4, 3, 3, 4, 3, 2, 3, 3], device='xla:1')
training step : 0 loss_reduced : 1.3349609375
tensor([[nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan],
        [nan, nan, nan, nan, nan]], device='xla:1', grad_fn=<AddmmBackward>) tensor([3, 1, 3, 3, 1, 3, 3, 0, 0, 4, 3, 3, 3, 3, 3, 3], device='xla:1')
training step : 1 loss_reduced : 0.0`
I also want to correct my mistake,the loss become zero at second mini-batch,rather than second epoch.

**Comment by taylanbil** [comment]
so, the input is not nan, but the output is nan? what is the loss value in the step *before* you see the nans?

**Comment by 1374839016** [comment]
Yes,input is correct.It's not nan,but the model output nan,so I think it's the problem related to model parameters.Before I see nan,the loss is normal,usually 1~2,so the first mini-batch doesn't have any issue.

**Comment by taylanbil** [comment]
can you check if model params are nan?
something like:
```python
for p in model.params():
  print(p.numel(), p.isnan().sum())
```

**Comment by 1374839016** [comment]
I check the parameters of my model,for 8 TPU cores,at first mini-batch,all model has normal parameters,they are not nan,but since second mini-batch,the model parameters become nan.And I also check the output for each TPU core,I find at first mini-batch,five cores can output correct output,others will output nan. I print the parameters of these model which output nan,their parameters aren't nan,but their output are nan.

**Comment by taylanbil** [comment]
I think we could focus on the one core case to debug.

We would need to edit the model's forward method to isolate when the flowing tensor becomes nan during the fwd pass. This is certainly not related to xm.optimizer_step, but it could be a bug in torch_xla.

Are you able to put some checks after each model layer, and see when it produces a nan?

**Comment by 1374839016** [comment]
Yes,you are right,the output become nan when pass through the network.The output of first few layers are normal,and become nan after several layers.
![image](https://user-images.githubusercontent.com/51019236/106869302-be00f600-670a-11eb-838c-a1295f9af18c.png)
And I also find this bug is related to model architecture,the model will have normal output with resnext50_32x4d archirecture,but nan with tf_efficientnet_b4_ns archetecture,both models are loaded from timm package.

**Comment by taylanbil** [comment]
good progress.. now we need to understand what kind of layer is "blocks", and what is the input to it.

**Comment by taylanbil** [comment]
I'm not experienced w/ Kaggle, I tried to repro but I'm getting
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:19 --:--:--     0     0curl: (6) Could not resolve host: raw.githubusercontent.com
python: can't open file 'pytorch-xla-env-setup.py': [Errno 2] No such file or directory
```

The curl command times out, looks like my notebook has no access to internet?

**Comment by 1374839016** [comment]
Yes,you must make sure the internet is connected before running the code. Internet button is on the right.

**Comment by taylanbil** [comment]
I'm not sure how to edit the model's forward in this environment, in order to inspect the input to `self.blocks` and the output.

**Comment by 1374839016** [comment]
Sorry，I rarely edit the forward directly，maybe you can find answer on google.
